include/deal.II/multigrid/mg_transfer_global_coarsening.h

```diff
@@ -0,0 +1,444 @@
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_mg_transfer_global_coarsening_h
+#define dealii_mg_transfer_global_coarsening_h
+
+#include <deal.II/base/mg_level_object.h>
+
+#include <deal.II/dofs/dof_handler.h>
+
+#include <deal.II/lac/affine_constraints.h>
+#include <deal.II/lac/la_parallel_vector.h>
+
+#include <deal.II/multigrid/mg_base.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+// Forward declarations
+#ifndef DOXYGEN
+namespace internal
+{
+  class MGTwoLevelTransferImplementation;
+}
+#endif
+
+
+
+/**
+ * Class for transfer between two multigrid levels for p or global coarsening.
+ */
+template <int dim, typename VectorType>
+class MGTwoLevelTransfer
+{
+public:
+  /**
+   * Perform prolongation.
+   */
+  void
+  prolongate(const unsigned int, VectorType &dst, const VectorType &src) const;
+
+  /**
+   * Perform restriction.
+   */
+  void
+  restrict_and_add(const unsigned int,
+                   VectorType &      dst,
+                   const VectorType &src) const;
+};
+
+
+
+/**
+ * Class for transfer between two multigrid levels for p or global coarsening.
+ * Specialization for LinearAlgebra::distributed::Vector.
+ */
+template <int dim, typename Number>
+class MGTwoLevelTransfer<dim, LinearAlgebra::distributed::Vector<Number>>
+{
+public:
+  /**
+   * Set up global coarsening transfer @p transfer between the given
+   * dof-handlers.
+   */
+  void
+  reinit_geometric_transfer(const DoFHandler<dim> &          dof_handler_fine,
+                            const DoFHandler<dim> &          dof_handler_coarse,
+                            const AffineConstraints<Number> &constraint_fine,
+                            const AffineConstraints<Number> &constraint_coarse);
+
+  /**
+   * Set up polynomial coarsening transfer @p transfer between the given
+   * dof-handlers.
+   *
+   * @note The function polynomial_transfer_supported() can be used to
+   *   check if the given polynomial coarsening strategy is supported.
+   */
+  void
+  reinit_polynomial_transfer(
+    const DoFHandler<dim> &          dof_handler_fine,
+    const DoFHandler<dim> &          dof_handler_coarse,
+    const AffineConstraints<Number> &constraint_fine,
+    const AffineConstraints<Number> &constraint_coarse);
+
+  /**
+   * Check if the polynomial transfer between @p fe_degree_fine and
+   * @p fe_degree_coarse is supported.
+   *
+   * @note Currently, the polynomial coarsening strategies: 1) go-to-one,
+   *   2) bisect, and 3) decrease-by-one is supported.
+   */
+  static bool
+  polynomial_transfer_supported(const unsigned int fe_degree_fine,
+                                const unsigned int fe_degree_coarse);
+
+  /**
+   * Perform prolongation.
+   */
+  void
+  prolongate(const unsigned int,
+             LinearAlgebra::distributed::Vector<Number> &      dst,
+             const LinearAlgebra::distributed::Vector<Number> &src) const;
```

Comment:
Since this does not adhere to the general interface `MGTransferBase`, do you think the `to_level` integer in `prolongate` is still needed? Below I see that this gets set to `invalid_unsigned_int`, which suggests to me that a two-level transfer should not need the index.